pallets/aggregate/src/lib.rs

```diff
@@ -0,0 +1,602 @@
+                            cause: CannotAggregateCause::InsufficientFound,
+                        });
+
+                        log::debug!("Failed to reserve balance {err:?}");
+                    })
+                else {
+                    return;
+                };
+
+                // We can add the statement and check if we should also move the aggregation in the should publish set
+                Self::deposit_event(Event::<T>::NewProof {
+                    statement,
+                    domain_id,
+                    aggregation_id: domain.next.id,
+                });
+                let to_publish = append_statement::<T>(domain, account.clone(), reserve, statement);
+                if let Some(aggregation) = to_publish {
+                    available_aggregation::<T>(domain, aggregation);
+                }
+            });
+        }
+    }
+
+    fn append_statement<T: Config>(
+        domain: &mut Domain<T>,
+        account: T::AccountId,
+        reserve: BalanceOf<T>,
+        statement: H256,
+    ) -> Option<Aggregation<T>> {
+        let aggregation = &mut domain.next;
+        aggregation
+            .statements
+            .force_push(StatementEntry::new(account.clone(), reserve, statement));
+        if aggregation.size as usize <= aggregation.statements.len() {
+            Some(sp_std::mem::replace(
+                aggregation,
+                aggregation.create_next(aggregation.size),
+            ))
+        } else {
+            None
+        }
+    }
+
+    fn available_aggregation<T: Config>(domain: &mut Domain<T>, aggregation: Aggregation<T>) {
+        Pallet::<T>::deposit_event(Event::<T>::ReadyToAggregate {
+            domain_id: domain.id,
+            aggregation_id: aggregation.id,
+        });
+        domain
+            .should_publish
+            .try_insert(aggregation.id, aggregation)
+            .expect("Should not publish aggregation if it's not possible: qed");
+        // If is full send an alert event
+        if domain.should_publish.len() >= domain.publish_queue_size as usize {
+            Pallet::<T>::deposit_event(Event::<T>::DomainFull {
+                domain_id: domain.id,
+            });
+        }
+    }
+
+    // Errors inform users that something went wrong.
+    #[pallet::error]
+    pub enum Error<T> {
+        /// This domain id is unknown.
+        UnknownDomainId,
+        /// This aggregation cannot be published or it's already published.
+        InvalidAggregationId,
+        /// The domain params are invalid.
+        InvalidDomainParams,
+    }
+
+    #[derive(Debug, Clone, PartialEq, Encode, Decode, TypeInfo, MaxEncodedLen)]
+    pub enum CannotAggregateCause {
+        NoAccount,
+        DomainNotRegistered { domain_id: u32 },
+        DomainStorageFull { domain_id: u32 },
+        InsufficientFound,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        NewDomain {
```

Comment:
I thought about that... maybe, and we can also think to remove `AggregationRemoved` one because we can relay on it. 
If we introduce it we have

Pros:
- A simpler `unregister_domain` with just one event
- A simpler `unregister_domain` weight (the one implemented now is buggy because it charge the costs about a full domain in the wrost case)
Cons:
- The user sould lissen this event instead the one about its own aggregation: is it possible to have some race condition on the client side that can lost an aggregation event?